File: C#/Socketry/socketry/Link.cs
using packetparser;
using socket;
using System.Collections.Concurrent;
using System.Net;

namespace socketry
{
    /// <summary>
    /// Class to create a new link.
    /// </summary>
    public class Link
    {
        ISocket _clientChannel;
        private MemoryStream _leftOverBuffer;

        public SocketPacket _currentSocketPacket;

        private ConcurrentQueue<Packet> _packets;

        /// <summary>
        /// Constructor for Link Class
        /// </summary>
        /// <param name="port">the port to start the link at</param>
        public Link(int port)
        {
            _clientChannel = new SocketTCP();
            _clientChannel.ConfigureBlocking(true); // block till connection is established
            _clientChannel.Connect(new IPEndPoint(IPAddress.Loopback, port));
            _clientChannel.ConfigureBlocking(false);
            _packets = new ConcurrentQueue<Packet>();
        }

        /// <summary>
        /// Constructor for Link class
        /// </summary>
        /// <param name="_connectedChannel">the socket to start the link on</param>
        public Link(ISocket _connectedChannel)
        {
            _clientChannel = _connectedChannel;
            _clientChannel.ConfigureBlocking(false);
            _packets = new ConcurrentQueue<Packet>();
        }

        /// <summary>
        /// Function to set the blocking state of the socket.
        /// </summary>
        /// <param name="to_block">the state to set to</param>
        public void ConfigureBlocking(bool to_block)
        {
            _clientChannel.ConfigureBlocking(to_block);
        }

        /// <summary>
        /// Function to register the socket to the selector.
        /// </summary>
        /// <param name="selector">The selector to register with</param>
        public void Register(Selector selector)
        {
            if (_clientChannel != null)
            {
                _clientChannel.Register(selector, SelectionKey.OP_READ, this);
            }
        }

        /// <summary>
        /// Function to read data from the sockets
        /// </summary>
        /// <returns>The stream containing the data read</returns>
        private MemoryStream ReadData()
        {
            int bufferLength = 1024;
            if (_leftOverBuffer != null)
            {
                bufferLength = Math.Max(1024, _leftOverBuffer.Capacity + 1024);
            }
            MemoryStream readBuffer = new MemoryStream(bufferLength);
            if (_leftOverBuffer != null)
            {
                readBuffer.Write(_leftOverBuffer.ToArray());
            }
            int dataRead = _clientChannel.Read(readBuffer);
            if (dataRead == 0)
            {
                // Raise runtime Exception
                Console.WriteLine("no data received...");
            }
            readBuffer.Position = 0;
            return readBuffer;
        }

        /// <summary>
        /// Function to get all the received packets.
        /// </summary>
        /// <returns>the list of all packets</returns>
        public List<Packet> GetPackets()
        {
            try
            {
                ReadAndParseAllPackets();
            }
            catch (IOException e)
            {
                Console.WriteLine(e.ToString());
                return new List<Packet>();
            }
            List<Packet> packets = new List<Packet>(_packets);
            _packets.Clear();
            return packets;
        }

        /// <summary>
        /// Function to read and parse all packets
        /// </summary>
        void ReadAndParseAllPackets()
        {
            bool initialState = _clientChannel.IsBlocking();
            int dataRead = ReadAndParsePackets();
            Console.WriteLine("Read packets...");
            _clientChannel.ConfigureBlocking(false);

            while (dataRead > 0)
            {
                dataRead = ReadAndParsePackets();
            }
            _clientChannel.ConfigureBlocking(initialState);
        }

        /// <summary>
        /// Function to get the length.
        /// </summary>
        /// <param name="data">the data to read</param>
        /// <param name="pos">the position of data</param>
        /// <returns></returns>
        static int ReadInt(byte[] data, int pos)
        {
            if (data.Length < pos)
            {
                throw new ArgumentException("Expected more data...");
            }
            return ((data[pos] & 0xFF) << 24) |
                ((data[pos + 1] & 0xFF) << 16) |
                ((data[pos + 2] & 0xFF) << 8) |
                (data[pos + 3] & 0xFF);
        }

        /// <summary>
        /// Function to parse packets
        /// </summary>
        /// <returns>the length of data parsed</returns>
        int ReadAndParsePackets()
        {
            MemoryStream buffer = ReadData();
            Console.WriteLine("Read from socket...");
            int remaining = (int)(buffer.Length - buffer.Position);
            byte[] data = new byte[remaining];
            buffer.Read(data, 0, remaining);

            int currDatapos = 0;
            int len = data.Length;
            while (currDatapos < len)
            {
                if (_currentSocketPacket == null)
                {
                    int left = len - currDatapos;
                    if (left < 4)
                    {
                        _leftOverBuffer = new MemoryStream(left);
                        _leftOverBuffer.Write(data, currDatapos, left);
                        _leftOverBuffer.Position = 0;
                        break;
                    }
                    int contentLength = ReadInt(data, currDatapos);
                    _currentSocketPacket = new SocketPacket(contentLength, new MemoryStream());
                    currDatapos += 4;
                }

                int toRead = Math.Min(_currentSocketPacket.BytesLeft, data.Length - currDatapos);
                _currentSocketPacket.Content.Write(data, currDatapos, toRead);
                _currentSocketPacket.BytesLeft -= toRead;
                currDatapos += toRead;

                if (_currentSocketPacket.BytesLeft == 0)
                {
                    _packets.Enqueue(Packet.Parse(_currentSocketPacket.Content.ToArray()));
                    _currentSocketPacket = null;
                }

            }
            return data.Length;
        }

        /// <summary>
        /// Function to get all packets
        /// </summary>
        /// <returns>The received packet</returns>
        public Packet GetPacket()
        {
            try
            {
                ReadAndParseAllPackets();
            }
            catch (IOException e)
            {
                Console.WriteLine(e.ToString());
                return null;
            }
            Packet packet = _packets.TryDequeue(out var p) ? p : null;
            Console.WriteLine($"packet {packet.ToString()}");
            return packet;
        }

        /// <summary>
        /// Function to send packet.
        /// </summary>
        /// <param name="packet">The packet to send to</param>
        /// <returns>The state of the packet</returns>
        public bool SendPacket(Packet packet)
        {
            byte[] packetData = Packet.Serialize(packet).ToArray();
            MemoryStream socketData = new MemoryStream(packetData.Length + 4);
            BinaryWriter writer = new BinaryWriter(socketData);
            writer.Write((byte)((packetData.Length >> 24) & 0xFF));
            writer.Write((byte)((packetData.Length >> 16) & 0xFF));
            writer.Write((byte)((packetData.Length >> 8) & 0xFF));
            writer.Write((byte)(packetData.Length & 0xFF));
            writer.Write(packetData);
            socketData.Position = 0;
            Console.WriteLine($"Sent packet {socketData.Length}");
            if (_clientChannel != null)
            {
                try
                {
                    _clientChannel.Write(socketData);
                }
                catch (IOException e)
                {
                    Console.WriteLine(e.ToString());
                }
            }
            return false;
        }
    }
}

File: JAVA/socketry-main/src/main/java/com/socketry/Link.java
package com.socketry;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.ClosedChannelException;
import java.nio.channels.SelectionKey;
import java.nio.channels.Selector;
import java.util.ArrayList;
import java.util.Queue;

import com.socketry.packetparser.Packet;
import com.socketry.socket.ISocket;
import com.socketry.socket.SocketTCP;

public class Link {
    private final ISocket clientChannel;
    private ByteBuffer leftOverBuffer;

    private SocketPacket currentSocketPacket;

    private final Queue<Packet> packets;

    /**
     * Blocks to connect to the given port
     *
     * @param _port server port to connect to.
     */
    public Link(int _port) throws IOException {
        clientChannel = new SocketTCP();
        clientChannel.configureBlocking(true); // block till connection is established
        clientChannel.connect(new InetSocketAddress(_port));
        clientChannel.configureBlocking(false);
        packets = new java.util.concurrent.ConcurrentLinkedQueue<>();
    }

    public Link(ISocket _connectedChannel) throws IOException {
        clientChannel = _connectedChannel;
        clientChannel.configureBlocking(false);
        packets = new java.util.concurrent.ConcurrentLinkedQueue<>();
    }

    public void configureBlocking(boolean to_block) throws IOException {
        clientChannel.configureBlocking(to_block);
    }

    public void register(Selector selector) throws ClosedChannelException {
        if (clientChannel != null) {
            clientChannel.register(selector, SelectionKey.OP_READ, this);
        }
    }

    /**
     * reads from the channel into the buffer
     * Puts the leftOverBuffer first then the data is read from the pipe
     * NOTE: Does **not** mark the leftOverBuffer as null. So it is responsibility
     * of callee to mark it null when utilized
     * throws Runtime Exception if not connected
     *
     * @return
     */
    private ByteBuffer readData() throws IOException {
        if (clientChannel == null) {
            // TODO : try to re-connect first
            throw new RuntimeException("Disconnected or never connected : ");
        }

        int bufferLength = 1024;
        if (leftOverBuffer != null) {
            bufferLength = Math.max(1024, leftOverBuffer.limit() + 1024);
        }
        ByteBuffer readBuffer = ByteBuffer.allocate(bufferLength);
        if (leftOverBuffer != null) {
            readBuffer.put(leftOverBuffer);
            // This moves the pointer inside the readBuffer
        }
        int dataRead = clientChannel.read(readBuffer);
//        System.out.println("readData : " + dataRead);
        if (dataRead == -1) {
            throw new RuntimeException("Disconnected or never connected : ");
        }
//        System.out.println("readBuffer.position() : " + readBuffer.position());
        readBuffer.flip();
        return readBuffer.slice();
    }

    /**
     * reads and returns packets received
     * Does not reentrant over the length of the array return
     * It could also be 0
     *
     * @return
     */
    public ArrayList<Packet> getPackets() {
        try {
            readAndParseAllPackets();
        } catch (IOException e) {
            System.out.println(e.getMessage());
            e.printStackTrace();
            return new ArrayList<>();
        }
        ArrayList<Packet> packetsToReturn = new ArrayList<>(packets);
        packets.clear();
        return packetsToReturn;
    }

    /**
     * Reads an int from the given data starting at the given position
     *
     * @param data the data to read from
     * @param pos  the position to start reading from
     * @return the int read
     */
    static int readInt(byte[] data, int pos) {
        if (data.length < pos) {
            throw new IllegalArgumentException("Data length is less than pos");
        }
        return ((data[pos] & 0xFF) << 24) |
            ((data[pos + 1] & 0xFF) << 16) |
            ((data[pos + 2] & 0xFF) << 8) |
            (data[pos + 3] & 0xFF);
    }

    void readAndParseAllPackets() throws IOException {
        boolean initialState = clientChannel.isBlocking();
        int dataRead = readAndParsePackets();
        clientChannel.configureBlocking(false);

        while (dataRead > 0) {
            dataRead = readAndParsePackets();
        }
        clientChannel.configureBlocking(initialState);
    }

    int readAndParsePackets() throws IOException {
        ByteBuffer buffer = readData();
        byte[] data = new byte[buffer.remaining()];
        buffer.get(data);

//        System.out.println("Read " + data.length + " bytes");

        int currDatapos = 0;
        int len = data.length;
        while (currDatapos < len) {
            if (currentSocketPacket == null) {
                // expecting a `Packet`
                int left = len - currDatapos;
                // System.out.println("LEFT OVER : " + left);
                if (left < 4) {
                    // got a leftover data store it for next parse.
                    leftOverBuffer = ByteBuffer.allocate(left);
                    leftOverBuffer.put(data, currDatapos, left);
                    leftOverBuffer.flip();
                    break;
                }
                int contentLength = readInt(data, currDatapos);
                // System.out.println("Content length : " + contentLength);
                currentSocketPacket = new SocketPacket(contentLength, new ByteArrayOutputStream());
                currDatapos += 4;
            }
            // check how much data is still to be read
            // and how much can be read
            int to_read = Math.min(currentSocketPacket.bytesLeft, data.length - currDatapos);
            // System.out.println("to_read : " + to_read + " currDatapos : " + currDatapos +
            // " bytesLeft : "
            // + currentSocketPacket.bytesLeft);
            currentSocketPacket.content.write(data, currDatapos, to_read);
            currentSocketPacket.bytesLeft -= to_read;
            currDatapos += to_read;
            // System.out.println("Bytes left : " + currentSocketPacket.bytesLeft + "
            // currDatapos : " + currDatapos);
            if (currentSocketPacket.bytesLeft == 0) {
                packets.add(Packet.parse(currentSocketPacket.content.toByteArray()));
                currentSocketPacket = null;
            }
        }
        return data.length;
    }

    public Packet getPacket() {
        try {
            readAndParseAllPackets();
        } catch (IOException e) {
            System.out.println(e.getMessage());
            e.printStackTrace();
            return null;
        }
        return packets.poll();
    }

    public boolean sendPacket(Packet packet) {
        byte[] packetData = Packet.serialize(packet).array();

        ByteBuffer socketData = ByteBuffer.allocate(packetData.length + 4);
        socketData.putInt(packetData.length);
        socketData.put(packetData);
        socketData.flip();
//        System.out.println("Sending " + packetData.length + " bytes");
//        System.out.println("clientChannel : " + clientChannel);Z

        if (clientChannel != null) {
            try {
                System.out.println("Packet sent "+packetData.length);
                clientChannel.write(socketData);
                return true;
            } catch (IOException e) {
                System.out.println("Error while writing" + e.getMessage());
                e.printStackTrace();
            }
        }
        return false;
    }
}

class SocketPacket {
    int bytesLeft;
    ByteArrayOutputStream content;

    public SocketPacket(int contentLength, ByteArrayOutputStream content) {
        this.bytesLeft = contentLength;
        this.content = content;
    }
}
